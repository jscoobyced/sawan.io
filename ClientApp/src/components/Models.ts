import { match } from "react-router-dom";

export interface SelectOption {
    value: number;
    text: string;
}

export interface IdParam {
    id: string;
}
