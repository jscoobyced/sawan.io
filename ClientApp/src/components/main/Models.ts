import { NavigationMenuContent } from "../../services/Models";

export interface HeaderProps {
    navigationMenuContent: NavigationMenuContent;
    signIn: (isSignedIn: boolean) => void;
    isSignedIn: boolean;
}
