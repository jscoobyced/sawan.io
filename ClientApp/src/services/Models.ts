export enum Language {
    English = 0,
    French = 1
}

export interface Link {
    text: string;
    url: string;
    target?: string;
    title?: string;
}

export interface MainContent {
    navigationMenuContent: NavigationMenuContent;
    menuContent: MenuContent;
    footerContent: FooterContent;
}

export interface NavigationMenuContent {
    home: string;
    websiteName: string;
    applications: string;
    cryptoCurrency: string;
    healthMonitor: string;
    about: string;
    information: string;
    resume: string;
}

export interface BlogElement {
    id: number;
    blogDate: Date;
    article: string;
    articleTitle: string;
}

export interface BlogPage {
    articles: BlogElement[];
}

export interface MenuContent {
    links: Link[];
}

export interface FooterContent {
    copyright: string;
    credits: string;
    year: string;
}
