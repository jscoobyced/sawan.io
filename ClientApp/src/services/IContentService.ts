import { BlogElement, BlogPage, Language, MainContent } from './Models';

export interface IContentService {
    getMainContent(language: Language): Promise<MainContent>;
    getDefaultMainContent(): MainContent;
    getBlogPage(maxResult: number): Promise<BlogPage>;
    getDefaultBlogPage(): BlogPage;
    getBlogElement(id: number): Promise<BlogElement>;
}
