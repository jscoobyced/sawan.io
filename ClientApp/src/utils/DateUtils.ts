export class DateUtil {
    public static toHumanHours(date: Date): string {
        return `${this.toDigits(date.getHours(), 2)}:${this.toDigits(date.getMinutes(), 2)}`;
    }

    public static dateToString(date: Date): string {
        const lDate = new Date(date);
        const year = lDate.getFullYear();
        const month = DateUtil.toDigits(lDate.getMonth() + 1, 2);
        const day = DateUtil.toDigits(lDate.getDate(), 2);
        return `${year}-${month}-${day}`;
    }

    public static dateToYyyyMmString(date: Date): string {
        const lDate = new Date(date);
        const year = lDate.getFullYear();
        const month = DateUtil.toDigits(lDate.getMonth() + 1, 2);
        return `${year}${month}`;
    }

    public static defaultDate(now = false): Date {
        if (now) {
            const nowDate = new Date();
            const month = DateUtil.toDigits(nowDate.getMonth() + 1, 2);
            const day = DateUtil.toDigits(nowDate.getDate(), 2);
            return new Date(`${nowDate.getFullYear()}-${month}-${day}T00:00:00Z`);
        }

        return new Date('2019-01-01T00:00:00Z');
    }

    private static toDigits(value: number, length: number): string {
        let result: string = value.toString();
        while (result.length < length) {
            result = "0" + result;
        }

        return result;
    }
}
