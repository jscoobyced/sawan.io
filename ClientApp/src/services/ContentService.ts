import { ApiUtils } from '../utils/ApiUtils';
import { IContentService } from './IContentService';
import { BlogElement, BlogPage, Language, MainContent } from './Models';

export class ContentService implements IContentService {

    public async getMainContent(language: Language): Promise<MainContent> {
        return this.fetchMainContent(language);
    }

    public getDefaultMainContent(): MainContent {
        return this.defaultMainContent();
    }

    public async getBlogPage(yearMonth: string, maxResult: number): Promise<BlogPage> {
        return this.updateBlogPage(yearMonth, maxResult);
    }

    public async getBlogElement(id: number): Promise<BlogElement> {
        return ApiUtils.fetchData<BlogElement>(`/api/Data/blog/${id}`);
    }

    public getDefaultBlogPage(): BlogPage {
        return {
            articles: []
        };
    }

    public async saveBlogElement(blogElement: BlogElement): Promise<boolean> {
        return this.doSaveBlogElement(blogElement);
    }

    protected async doSaveBlogElement(blogElement: BlogElement): Promise<boolean> {
        return ApiUtils.postData('/api/Data/blog/post/', blogElement);
    }

    protected async fetchMainContent(language: Language): Promise<MainContent> {
        return ApiUtils.fetchData<MainContent>(`/api/Data/main/${language}`);
    }

    protected defaultMainContent(): MainContent {
        return {
            menuContent: {
                title: '',
                links: []
            },
            navigationMenuContent: {
                about: '',
                applications: '',
                cryptoCurrency: '',
                healthMonitor: '',
                home: '',
                websiteName: '',
                information: '',
                resume: ''
            },
            footerContent: {
                copyright: '',
                credits: '',
                year: ''
            }
        };
    }

    protected async updateBlogPage(yearMonth: string, maxResult: number): Promise<BlogPage> {
        const empty = {
            articles: []
        };
        if (maxResult <= 0) {
            return empty;
        }
        const data = await ApiUtils.fetchData<BlogElement[]>(`/api/Data/blogpage/${yearMonth}/${maxResult}`);
        if (!data || data.length === 0) {
            return empty;
        }
        return {
            articles: data.slice(0, Math.min(maxResult, data.length))
        };
    }
}
