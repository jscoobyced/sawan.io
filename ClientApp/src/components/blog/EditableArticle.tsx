import React from 'react';
import { BlogElement } from '../../services/Models';
import { DateUtil } from '../../utils/DateUtils';
import { Article } from './Article';

export interface EditableArticleProps {
    blogElement: BlogElement;
    update: (article: string, articleTitle: string, articleBlogDate: Date, id: number) => Promise<boolean>;
}

export interface EditableArticleState {
    blogElement: BlogElement;
    saveText: string;
    saveDisabled: boolean;
}

export class EditableArticle extends React.Component<EditableArticleProps, EditableArticleState> {
    private readonly SAVE = 'Save';
    private readonly SAVING = 'Saving';
    private readonly SAVED = 'Saved';
    private readonly FAILED = 'Failed';
    public constructor(props: EditableArticleProps) {
        super(props);
        this.state = {
            blogElement: props.blogElement,
            saveText: this.SAVE,
            saveDisabled: false
        };
    }

    public componentDidUpdate(prevProps: EditableArticleProps) {
        const blogElement = this.props.blogElement;
        if (blogElement.article !== prevProps.blogElement.article) {
            this.setState({
                blogElement
            });
        }
    }

    public render() {
        if (!this.state.blogElement) {
            return <></>;
        }

        const { article, articleTitle } = this.state.blogElement;
        return (
            <div>
                <article
                    className='blog-article'
                >
                    <input
                        type='text'
                        id='article-title'
                        value={articleTitle}
                        onChange={this.onChangeArticleTitle}
                    />
                    <textarea
                        rows={24}
                        id='article-text'
                        value={article}
                        onChange={this.onChangeArticle}
                    />
                    <button
                        className='btn-article'
                        id='article-save'
                        onClick={this.update}
                        disabled={this.state.saveDisabled}
                    >
                        {this.state.saveText}
                    </button>
                </article>
                <Article
                    blogElement={this.state.blogElement}
                    backLink={true}
                    ellipsis={false}
                />
            </div>);
    }

    private readonly onChangeArticle = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
        event.preventDefault();
        const blogElement = { ...this.state.blogElement, article: event.target.value };
        this.setState({
            blogElement,
            saveText: this.SAVE,
            saveDisabled: false
        });
    }

    private readonly onChangeArticleTitle = (event: React.ChangeEvent<HTMLInputElement>) => {
        event.preventDefault();
        const blogElement = { ...this.state.blogElement, articleTitle: event.target.value };
        this.setState({
            blogElement,
            saveText: this.SAVE,
            saveDisabled: false
        });
    }

    private readonly update = (event: React.MouseEvent<HTMLButtonElement>) => {
        event.preventDefault();
        this.setState({
            saveText: this.SAVING,
            saveDisabled: true
        });
        const { article, articleTitle, blogDate, id } = this.state.blogElement;
        this.props.update(article, articleTitle, blogDate, id)
            .then(success => {
                this.setState({
                    saveText: success ? this.SAVED : this.FAILED
                });
            });
    }
}
