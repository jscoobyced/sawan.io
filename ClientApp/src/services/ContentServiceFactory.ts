import { ContentService } from './ContentService';
import { IContentService } from './IContentService';
import { MockContentService } from './MockContentService';

export class ContentServiceFactory {

    public static GetContentService(): IContentService {
        const mode = process.env.mode as string;
        if (mode === 'production') {
            return new ContentService();
        }

        return new MockContentService();
    }
}
