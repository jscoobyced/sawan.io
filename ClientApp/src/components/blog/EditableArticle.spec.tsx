import { shallow } from 'enzyme';
import React from 'react';
import { BlogElement } from '../../services/Models';
import { DateUtil } from '../../utils/DateUtils';
import { EditableArticle } from './EditableArticle';

test('Render EditableArticle', () => {
    const update = jest.fn().mockImplementation();
    const editableArticle = shallow((
        <EditableArticle
            blogElement={null as unknown as BlogElement}
            update={update}
        />));
    expect(editableArticle).toMatchSnapshot();
});

test('Render EditableArticle', () => {
    const update = jest.fn().mockImplementation();
    const blogElement: BlogElement = {
        article: 'test',
        articleTitle: 'blablabla',
        id: '123',
        blogDate: DateUtil.defaultDate()
    };
    const editableArticle = shallow((
        <EditableArticle
            blogElement={blogElement}
            update={update}
        />));
    expect(editableArticle).toMatchSnapshot();
});

test('Render EditableArticle', () => {
    const update = jest.fn().mockImplementation((): Promise<boolean> => {
        return Promise.resolve(true);
    });
    const blogElement: BlogElement = {
        article: 'test',
        articleTitle: 'blablabla',
        id: '123',
        blogDate: DateUtil.defaultDate()
    };
    const editableArticle = shallow((
        <EditableArticle
            blogElement={blogElement}
            update={update}
        />));
    editableArticle
        .find('#article-title')
        .simulate('change', { target: { value: 'bla-title' }, preventDefault: () => { return; } });
    editableArticle
        .find('#article-text')
        .simulate('change', { target: { value: 'bla-text' }, preventDefault: () => { return; } });
    editableArticle
        .find('#article-save')
        .simulate('click', { preventDefault: () => { return; } });
    expect(editableArticle).toMatchSnapshot();
    expect(update).toHaveBeenCalledTimes(1);
});
