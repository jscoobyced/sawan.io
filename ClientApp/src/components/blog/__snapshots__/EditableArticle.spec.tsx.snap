// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render EditableArticle 1`] = `
<div>
  <article
    className="blog-article"
  >
    <input
      id="article-title"
      onChange={[Function]}
      type="text"
      value=""
    />
    <textarea
      id="article-text"
      onChange={[Function]}
      rows={24}
      value=""
    />
    <button
      className="btn-article"
      disabled={false}
      id="article-save"
      onClick={[Function]}
    >
      Save
    </button>
  </article>
  <Article
    backLink={true}
    blogElement={
      Object {
        "article": "",
        "articleTitle": "",
        "blogDate": 2019-03-14T00:00:00.000Z,
        "id": "",
        "updateDate": 2019-01-01T00:00:00.000Z,
      }
    }
    ellipsis={false}
  />
</div>
`;

exports[`Render EditableArticle 2`] = `
<div>
  <article
    className="blog-article"
  >
    <input
      id="article-title"
      onChange={[Function]}
      type="text"
      value=""
    />
    <textarea
      id="article-text"
      onChange={[Function]}
      rows={24}
      value=""
    />
    <button
      className="btn-article"
      disabled={false}
      id="article-save"
      onClick={[Function]}
    >
      Save
    </button>
  </article>
  <Article
    backLink={true}
    blogElement={
      Object {
        "article": "",
        "articleTitle": "",
        "blogDate": 2019-03-14T00:00:00.000Z,
        "id": "",
        "updateDate": 2019-01-01T00:00:00.000Z,
      }
    }
    ellipsis={false}
  />
</div>
`;

exports[`Render EditableArticle 3`] = `
<div>
  <article
    className="blog-article"
  >
    <input
      id="article-title"
      onChange={[Function]}
      type="text"
      value="bla-title"
    />
    <textarea
      id="article-text"
      onChange={[Function]}
      rows={24}
      value="bla-text"
    />
    <button
      className="btn-article"
      disabled={true}
      id="article-save"
      onClick={[Function]}
    >
      Saving
    </button>
  </article>
  <Article
    backLink={true}
    blogElement={
      Object {
        "article": "bla-text",
        "articleTitle": "bla-title",
        "blogDate": 2019-03-14T00:00:00.000Z,
        "id": "",
        "updateDate": 2019-01-01T00:00:00.000Z,
      }
    }
    ellipsis={false}
  />
</div>
`;
